### ISSUE TITLE
Doesn't work with VTK 9

### ISSUE THREAD

**Issue by embeddedmz** [embeddedmz/ParaViewScalarBarWidget#1]
Doesn't compile with VTK9, even if you manage to fix compile errors, there will be a null pointer that will crash the app.

Solution : update the ParaView files.

**Comment by PauloCarvalhoRJ** [comment]
Hello, I'm using `vtkParavewScalarBar` in my project.  If it happens to me and find a solution, I'll share with the community.  For now, I'm working with VTK 8.2 with no plan to switch to VTK 9 soon.

**Comment by PauloCarvalhoRJ** [comment]
Cross referencing to share a potential solution: https://github.com/PauloCarvalhoRJ/gammaray/issues/262

**Comment by embeddedmz** [comment]
Hi @PauloCarvalhoRJ
VS2019 showed me errors in vtkParaViewScalarBar.cxx, there are missing return statements in : 
```
//----------------------------------------------------------------------------
const std::string& vtkParaViewScalarBar::GetTitle()
{
    (this->ScalarBarActor->GetTitle()) ? this->ScalarBarActor->GetTitle() : "";
}

//----------------------------------------------------------------------------
void vtkParaViewScalarBar::SetComponentTitle(const std::string& title)
{
    this->ScalarBarActor->SetComponentTitle(title.c_str());
}

//----------------------------------------------------------------------------
const std::string& vtkParaViewScalarBar::GetComponentTitle()
{
    (this->ScalarBarActor->GetComponentTitle()) ? this->ScalarBarActor->GetComponentTitle() : "";
}
```

For historical reasons, this is not considered errors in other compilers. I'm glad that VS2019 consider them as errors !

Best regards.

**Comment by embeddedmz** [comment]
By the way, I compiled with the VTK provided by vcpkg.

**Comment by PauloCarvalhoRJ** [comment]
> 
> 
> Hi @PauloCarvalhoRJ
> VS2019 showed me errors in vtkParaViewScalarBar.cxx, there are missing return statements in :
> 
> ```
> //----------------------------------------------------------------------------
> const std::string& vtkParaViewScalarBar::GetTitle()
> {
>     (this->ScalarBarActor->GetTitle()) ? this->ScalarBarActor->GetTitle() : "";
> }
> 
> //----------------------------------------------------------------------------
> void vtkParaViewScalarBar::SetComponentTitle(const std::string& title)
> {
>     this->ScalarBarActor->SetComponentTitle(title.c_str());
> }
> 
> //----------------------------------------------------------------------------
> const std::string& vtkParaViewScalarBar::GetComponentTitle()
> {
>     (this->ScalarBarActor->GetComponentTitle()) ? this->ScalarBarActor->GetComponentTitle() : "";
> }
> ```
> 
> For historical reasons, this is not considered errors in other compilers. I'm glad that VS2019 consider them as errors !
> 
> Best regards.

Hello, I compile with MinGW64 in Windows and it treats missing returns as warnings.  In my case, missing returns for functions returning non-void result almost always in mysterious, random crashes.  I had to hunt them down in all my code to get rid of them.

**Comment by embeddedmz** [comment]
@PauloCarvalhoRJ There's another issue in the code, returing a temporary local reference. I just learned that today via a warning of an old version of Viusal Studio. I always thought that returning a const ref extends the lifetime of the temporary returned value :( Damn C++ !

The fix is to return a std::string. Otherwise, the methods are unsafe.

**Comment by PauloCarvalhoRJ** [comment]
Hi, @embeddedmz , well, a reference becomes invalid the moment the variable it refers to goes out of scope.  That's one of the reasons they invented move semantics. It allows the passing of variables/objects between scopes without copying them or resorting to pointers (smart or crude).

**Comment by AhmedShawi** [comment]
Hi, I am trying to compile on VTK 9 and I get these errors:
![image](https://user-images.githubusercontent.com/40696191/133063535-a8457065-b197-4c1c-87ca-a954129c9ca4.png)

There is also a missing ; before the return on line 110 in vtkBoundingRectContextDevice2D.cxx

Could you help resolve?

I've tried to compile using 1. QtCreator & qmake, 2. QtCreator & cmake and 3. VS2019 & cmake

**Comment by embeddedmz** [comment]
@AhmedShawi OK I will take a look at this afternoon.

You can get rid off vtkDiscretizableColorTransferFunction.cxx that is used in the Example folder (and modify the CMakeLists.txt). If I remember correctly, I added it because I was using an old VTK version and this class was outdated, so I override it by compiling it (I know it's something nasty).

**Comment by AhmedShawi** [comment]
Thank you, much appreciated!

After removing the vtkDiscretizableColorTransferFunction.cxx and .h, the error moves to the vtkContext2DScalarBarActor.cxx:
![image](https://user-images.githubusercontent.com/40696191/133068027-e54e77ed-ed3b-404b-87b2-8529fbde22d2.png)

ctf->MapValue(value) gives a 'const unsigned char *' but you can't use that to initialise an 'unsigned char *'. Could this be why you overrided vtkDiscretizableColorTransferFunction?

**Comment by embeddedmz** [comment]
@AhmedShawi, I don't remember, but in an older VTK version, there's a bug that make the computation of color slower because a certain method was not overriden in a subclass. Moreover, vtkDiscretizableColorTransferFunction is not available in older VTK versions. Without this class, computation of color take so much time resulting in slow renderings.

**Comment by PauloCarvalhoRJ** [comment]
> 
> 
> Thank you, much appreciated!
> 
> After removing the vtkDiscretizableColorTransferFunction.cxx and .h, the error moves to the vtkContext2DScalarBarActor.cxx:
> ![image](https://user-images.githubusercontent.com/40696191/133068027-e54e77ed-ed3b-404b-87b2-8529fbde22d2.png)
> 
> ctf->MapValue(value) gives a 'const unsigned char *' but you can't use that to initialise an 'unsigned char *'. Could this be why you overrided vtkDiscretizableColorTransferFunction?

It seems to me that `vtkContext2DScalarBarActor.cxx` must be tweaked to account for the newly added constness.  The VTK developers are gradually adding constness to the older parts of VTK code.

**Comment by AhmedShawi** [comment]
> It seems to me that `vtkContext2DScalarBarActor.cxx` must be tweaked to account for the newly added constness. The VTK developers are gradually adding constness to the older parts of VTK code.

I've just been trying to tweak it but can't seem to get it to work. 
I can get it to compile by changing line 282 in vtkContext2DScalarBarActor.cxx to:
        unsigned char color[3] = { static_cast<unsigned char>(255.0 * ctf->GetOpacity(originalValue) + 0.5) };
and commenting out line 278.

But doing so just throws an exception error in vtkpen:
![image](https://user-images.githubusercontent.com/40696191/133108931-c1d96aab-776a-46f3-b0f1-a1d150b56cd8.png)

Any chance you could have a look?

@embeddedmz If you manage to find a solution that would be very much appreciated.

**Comment by PauloCarvalhoRJ** [comment]
@AhmedShawi , Take a look at the message.  It says "this was nullptr".  It's saying that there was an attempt to access the `Color` member variable of a non-existent object.  I suggest that you build the application in debug mode (also use the VTK libraries compiled in debug mode) and repeat the test starting the application via the debugger.  Configure your kit in Qt Creator to use CBD that comes with the Windows Kit, which is a separate installation from Visual Studio.  Then hit "Run with debugger" button in Qt Creator.  It'll stop at the offending line showing the stack trace in a pane.  Use that to track down the origin of the null pointer.

If you are already running in debug mode, then it would be good to post the stack trace here.

**Comment by AhmedShawi** [comment]
So I have done the following:
1. Removed the included vtkDiscretizableColorTransferFunction.cxx and .h and removed it from the cmakelists.txt
2. Changed line 282 in vtkContext2DScalarBarActor.cxx to:
    `unsigned char color[3] = { static_cast(255.0 * ctf->GetOpacity(originalValue) + 0.5) };`
3. Commented out line 278 in vtkContext2DScalarBarActor.cxx

As mentioned, it breaks as noted above with the exception in vtkpen with **this** being a nullptr.

So I've done some debugging and found that the issue occurs when trying to render the scalarbar (renderer is set in line 94 in ParaViewScalarBar.cxx):
    `scalarBar->SetRenderer(renderer);`

Commenting out that line and running the example works fine (but of course there is no scalarbar):
![image](https://user-images.githubusercontent.com/40696191/133152179-ea2d86b9-060a-4697-b726-d06e2f1c5c9f.png)

Here is the stack trace if it helps:
![image](https://user-images.githubusercontent.com/40696191/133151845-c45174ce-feeb-4128-ab1d-3cbcc020d967.png)

I am trying to run the example given in this rep before applying elsewhere.

Any ideas?

**Comment by PauloCarvalhoRJ** [comment]
I assume `ParaViewScalarBar.exe` is your program.  Something in your program is triggering the crash within a VTK DLL.  You have to inspect the lines in the stack trace that refers to your code (those with a leading `ParaViewScalarBar.exe:`).

What is the content of the second line of the stack trace?  You have to go up the calling chain and inspect the values until you find which code is creating a null-pointer `vtkPen` object.

**Comment by AhmedShawi** [comment]
Thanks for the help.

Yes ParaViewScalarBar.exe is my program. I am just trying to run the example code in this repo https://github.com/embeddedmz/ParaViewScalarBarWidget

I have not added or amended anything from the example - only the three modifications listed previously.

I have found where the null-pointer pen is created:
![image](https://user-images.githubusercontent.com/40696191/133155483-fe6017af-d266-4ce1-9d40-1457c364142b.png)

Edit: I'm pretty sure this just links back to the changes I made re: the 'const's.

> Changed line 282 in vtkContext2DScalarBarActor.cxx to:
>    unsigned char color[3] = { static_cast(255.0 * ctf->GetOpacity(originalValue) + 0.5) };
> Commented out line 278 in vtkContext2DScalarBarActor.cxx

**Comment by PauloCarvalhoRJ** [comment]
Make sure you have a `vtkContext2D::Begin(vtkContextDevice2D *device))` ((https://github.com/Kitware/VTK/blob/db5498e8b45e8a1c97ef3856e46231c333c5893b/Rendering/Context2D/vtkContext2D.cxx#L38)) being called at least once *before* calling `getPen()` .  `Begin()` sets the `Device` member variable, which is used in `getPen()` to return the device's pen: https://github.com/Kitware/VTK/blob/db5498e8b45e8a1c97ef3856e46231c333c5893b/Rendering/Context2D/vtkContext2D.cxx#L905.

**Comment by PauloCarvalhoRJ** [comment]
I forgot to mention that at least one call to `Begin()` must be done before on the *same* object pointed by the local variable `painter` shown in your last screen capture.

**Comment by PauloCarvalhoRJ** [comment]
Maybe you should go further up (3rd line and on) in the stack trace trying to locate the code that is calling `Begin()` before calling `vtkContext2DScalarBarActor::Paint()`.

**Comment by AhmedShawi** [comment]
So I have found it. In vtkBoundingRectContextDevice2D.cxx (included in this repo), `Begin()` is called on `context` (line60) before it is passed into `Paint()`. Begin() is also called on `bbDevice` as you can see in the screenshot below:

![image](https://user-images.githubusercontent.com/40696191/133277469-ee9cd6c6-be88-41b6-9951-fe9bd87c4dd9.png)

It looks fine :(

@embeddedmz Did you manage to successfully get this working?

**Comment by PauloCarvalhoRJ** [comment]
It doesn't look ok for me.  You are calling `Paint()` on `item`.  Where is the prior call to `item->Begin()`?

**Comment by embeddedmz** [comment]
@embeddedmz Did you manage to successfully get this working? Nope, I gave up on VTK9 :p

**Comment by AhmedShawi** [comment]
I have followed the call stack backwards:

`item` is `this->ScalarBarItem`
![image](https://user-images.githubusercontent.com/40696191/133442884-13bd76d4-e9b9-4c90-9dd4-7cbbe618413c.png)

`item` is the `vtkContext2DScalarBarActor` (`this->GetScalarBarActor()`)
![image](https://user-images.githubusercontent.com/40696191/133443446-7af71201-336f-46f2-a7ad-a615ff1b9594.png)

@PauloCarvalhoRJ I think I'm a bit out of my depth here. Are you able to attempt to replicate? Thanks again for the help.

@embeddedmz do you know if this rep works and compiles well with VTK8.2.0? I may end up removing VTK9.0 and just using VTK8.2.0.

**Comment by embeddedmz** [comment]
@AhmedShawi If I remember correctly I think the answer is yes (with 8.0.x definitely yes). I'm using vcpkg on my Windows computers and I'm unable to build VTK properly as vcpkg (C++ library manager) is not playing well with Visual Studio (vcpkg integration is enabled in VS) and VTK source files, so I can't give you a clear answer right now !

**Comment by AhmedShawi** [comment]
> @AhmedShawi If I remember correctly I think the answer is yes (with 8.0.x definitely yes). I'm using vcpkg on my Windows computers and I'm unable to build VTK properly as vcpkg (C++ library manager) is not playing well with Visual Studio (vcpkg integration is enabled in VS) and VTK source files, so I can't give you a clear answer right now !

Thanks. I'll leave a VTK downgrade as a last resort, as I have written a LOT using VTK 9. Hopefully @PauloCarvalhoRJ will be able to help a little further :)

**Comment by PauloCarvalhoRJ** [comment]
> 
> 
> I have followed the call stack backwards:
> 
> `item` is `this->ScalarBarItem`
> ![image](https://user-images.githubusercontent.com/40696191/133442884-13bd76d4-e9b9-4c90-9dd4-7cbbe618413c.png)
> 
> `item` is the `vtkContext2DScalarBarActor` (`this->GetScalarBarActor()`)
> ![image](https://user-images.githubusercontent.com/40696191/133443446-7af71201-336f-46f2-a7ad-a615ff1b9594.png)
> 
> @PauloCarvalhoRJ I think I'm a bit out of my depth here. Are you able to attempt to replicate? Thanks again for the help.
> 
> @embeddedmz do you know if this rep works and compiles well with VTK8.2.0? I may end up removing VTK9.0 and just using VTK8.2.0.

Hello, @AhmedShawi.  Sorry for the late reply.  Sorry, no, I haven't moved to VTK9.0 yet.  I'll be able to hunt that bug when I do.  But, there is no shame in downgrading to a previous version if it comes to that.  I do this from time to time in my job.  Do you have any special reason to move to VTK 9 other than just have the latest version?

**Comment by PauloCarvalhoRJ** [comment]
> 
> 
> > @AhmedShawi If I remember correctly I think the answer is yes (with 8.0.x definitely yes). I'm using vcpkg on my Windows computers and I'm unable to build VTK properly as vcpkg (C++ library manager) is not playing well with Visual Studio (vcpkg integration is enabled in VS) and VTK source files, so I can't give you a clear answer right now !
> 
> Thanks. I'll leave a VTK downgrade as a last resort, as I have written a LOT using VTK 9. Hopefully @PauloCarvalhoRJ will be able to help a little further :)

As I said before: you have to make sure there is a call to `item->Begin()` or to whatever object `item` points to (`this->ScalarBarItem` or `actor`).  `actor` doesn't seem the right one.  Why don't you simple shove an `item->Begin()` before the `item->Paint(context.Get())` call that is abending?  See what happens...  Your computer won't blow up! 😃

**Comment by AhmedShawi** [comment]
> > > @AhmedShawi If I remember correctly I think the answer is yes (with 8.0.x definitely yes). I'm using vcpkg on my Windows computers and I'm unable to build VTK properly as vcpkg (C++ library manager) is not playing well with Visual Studio (vcpkg integration is enabled in VS) and VTK source files, so I can't give you a clear answer right now !
> > 
> > 
> > Thanks. I'll leave a VTK downgrade as a last resort, as I have written a LOT using VTK 9. Hopefully @PauloCarvalhoRJ will be able to help a little further :)
> 
> As I said before: you have to make sure there is a call to `item->Begin()` or to whatever object `item` points to (`this->ScalarBarItem` or `actor`). `actor` doesn't seem the right one. Why don't you simple shove an `item->Begin()` before the `item->Paint(context.Get())` call that is abending? See what happens... Your computer won't blow up! 😃

Thank you for getting back to me. Yeah I tried calling `item->Begin()` but there was no `Begin` within `item`. I have since discovered that it is the viewport which is NULL as well. After calling RenderWindow->Render(), the code goes through a few DLLs and ends up here, and this is where `viewport` first appears:
![image](https://user-images.githubusercontent.com/40696191/134316664-f37e1f67-4822-485e-b64c-59aa6c8ce880.png)

**Comment by PauloCarvalhoRJ** [comment]
True... `item` is a pointer to `vtkAbstractContextItem` which doesn't define a `Begin()`.  I got confused at come point.  You have to call `Begin()` on the `vtkContext2D` object, which is pointed to by `painter` in this code:  https://user-images.githubusercontent.com/40696191/133155483-fe6017af-d266-4ce1-9d40-1457c364142b.png .

**Comment by PauloCarvalhoRJ** [comment]
I was able to make it to work in my project (with VTK 9.1):
![image](https://user-images.githubusercontent.com/6398896/156846339-9cb6dca3-59d1-4d87-bc20-1cea7bd03514.png)
It looks small due to 4K issue, but the widget is working as expected.

Here's the upgraded code (not merged to master yet): https://github.com/PauloCarvalhoRJ/gammaray/tree/Upgrade_to_VTK9.1/imagejockey/paraviewscalarbar

**Comment by embeddedmz** [comment]
Hi @PauloCarvalhoRJ 

Great ! If you have time, you can make a pull request so other people using VTK can profit.

Best regards.

**Comment by PauloCarvalhoRJ** [comment]
Hi, @embeddedmz ,

I'm getting 403 errors when pushing.  I think you need to enable some kind of permission so I can push commits and create pull requests.

regards,

Paulo

**Comment by PauloCarvalhoRJ** [comment]
Hello, @embeddedmz 

I had to do it via a fork.  Here you are: https://github.com/embeddedmz/ParaViewScalarBarWidget/pull/2 .

all the best,

Paulo
